core/vm/zk_batch_counters.go

```diff
@@ -0,0 +1,130 @@
+package vm
+
+import (
+	"bytes"
+	"math"
+	"strconv"
+)
+
+type BatchCounterCollector struct {
+	// every new transaction added to the batch affects these counters, so they need their own entry as this will
+	// be overwritten time and again as transactions are added
+	l2DataCollector *CounterCollector
+	transactions    []*TransactionCounter
+	smtLevels       int
+	blockCount      int
+}
+
+func NewBatchCounterCollector(smtMaxLevel uint32) *BatchCounterCollector {
+	totalLevel := len(strconv.FormatInt(int64(math.Pow(2, float64(smtMaxLevel))+250000), 2))
+	return &BatchCounterCollector{
+		transactions: []*TransactionCounter{},
+		smtLevels:    totalLevel,
+		blockCount:   0,
+	}
+}
+
+// AddNewTransactionCounters makes the collector aware that a new transaction is attempting to be added to the collector
+// here we check the batchL2Data length and ensure that it doesn't cause an overflow.  This will be re-calculated
+// every time a new transaction is added as it needs to take into account all the transactions in a batch.
+// The function will return false in the case of an error or an overflow of counters
+func (bcc *BatchCounterCollector) AddNewTransactionCounters(txCounters *TransactionCounter) (bool, error) {
+	bcc.transactions = append(bcc.transactions, txCounters)
```

Comment:
hmm, what happens if it overflows? bcc.transactions stays the same?

Reply:
here and everywhere -- won't there be an extra operation?

Reply:
So the thought is that we only care for counters until they overflow, once it has or we reach the time limit we don't need to hold on to them.  So although the transactions slice will contain the transaction that caused the overflow it doesn't matter because we won't be including that last transaction in the block

Reply:
If you check in the stage_execution code you'll see we add a tx where this is called, then check for an overflow immediately, then execute, then check for an overflow post execution.  Only if it hasn't overflowed at these stages do we add the TX into the slice ready to be included in the block. If there's an overflow we quit early and process the block without this TX

Reply:
okay got it